packages/ai-jsx/src/core/log.ts

```diff
@@ -1,26 +1,61 @@
 import _ from 'lodash';
 import { pino } from 'pino';
 import { Element } from '../index';
-import { v4 as uuidv4 } from 'uuid';
 
 export type LogLevel = 'fatal' | 'error' | 'warn' | 'info' | 'debug' | 'trace';
 
 export type Logger = Record<LogLevel, (obj: object | string, msg?: string) => void>;
 
-/**
- * @param level The logging level.
- * @param element The element from which the log originated.
- * @param renderId A unique identifier associated with the rendering request.
- * @param metadataOrMessage An object to be included in the log, or a message to log.
- * @param message The message to log, if `metadataOrMessage` is an object.
- */
-export type LogImplementation = (
-  level: LogLevel,
-  element: Element<object>,
-  renderId: string,
-  metadataOrMessage: object | string,
-  message?: string
-) => void;
+export abstract class LogImplementation {
+  protected readonly loggedExceptions = new WeakMap<object, boolean>();
+
+  /**
+   * @param level The logging level.
+   * @param element The element from which the log originated.
+   * @param renderId A unique identifier associated with the rendering request.
```

Comment:
This comment should say whether `renderId` is scoped to the entire top-level render or just the individual component render.